File: ts/7 - Trees/427.ConstructQuadTree.ts
/* 
Given a n * n matrix grid of 0's and 1's only. We want to represent grid with a Quad-Tree.

Return the root of the Quad-Tree representing grid.

A Quad-Tree is a tree data structure in which each internal node has exactly four children. Besides, each node has two attributes:

val: True if the node represents a grid of 1's or False if the node represents a grid of 0's. Notice that you can assign the val to True or False when isLeaf is False, and both are accepted in the answer.
isLeaf: True if the node is a leaf node on the tree or False if the node has four children.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;
}
We can construct a Quad-Tree from a two-dimensional area using the following steps:

If the current grid has the same value (i.e all 1's or all 0's) set isLeaf True and set val to the value of the grid and set the four children to Null and stop.
If the current grid has different values, set isLeaf to False and set val to any value and divide the current grid into four sub-grids as shown in the photo.
Recurse for each of the children with the proper sub-grid.
*/

/* ------------------------------------------------------------------------------- */

class Node {
    val: boolean;
    isLeaf: boolean;
    topLeft: Node | null;
    topRight: Node | null;
    bottomLeft: Node | null;
    bottomRight: Node | null;
    constructor(
        val?: boolean,
        isLeaf?: boolean,
        topLeft?: Node,
        topRight?: Node,
        bottomLeft?: Node,
        bottomRight?: Node
    ) {
        this.val = val === undefined ? false : val;
        this.isLeaf = isLeaf === undefined ? false : isLeaf;
        this.topLeft = topLeft === undefined ? null : topLeft;
        this.topRight = topRight === undefined ? null : topRight;
        this.bottomLeft = bottomLeft === undefined ? null : bottomLeft;
        this.bottomRight = bottomRight === undefined ? null : bottomRight;
    }
}

function construct(grid: number[][]): Node | null {
    function dfs(n: number, r: number, c: number) {
        let allSame = true;
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (grid[r][c] !== grid[r + i][c + j]) {
                    allSame = false;
                    break;
                }
            }
        }

        if (allSame) {
            return new Node(!!grid[r][c], true);
        }

        n = Math.floor(n / 2);

        const topLeft = dfs(n, r, c);
        const topRight = dfs(n, r, c + n);
        const bottomLeft = dfs(n, r + n, c);
        const bottomRight = dfs(n, r + n, c + n);

        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }

    return dfs(grid.length, 0, 0);
}

/* ------------------------------------------------------------------------------- */

/*
T.C.: O(N^2log(N))
S.C.: O(log(N))
*/

File: python/7 - Trees/427.ConstructQuadTree.py
# Definition for a QuadTree node.
from typing import List


class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight


# class Solution:
#     def construct(self, grid: List[List[int]]) -> 'Node':
#         def dfs(n: int, r: int, c: int) -> Node:
#             isAllSame = True
#             for i in range(n):
#                 for j in range(n):
#                     if grid[r][c] != grid[r + i][c + j]:
#                         isAllSame = False
#                         break

#             if isAllSame:
#                 return Node(grid[r][c], True, None, None, None, None)

#             n = n // 2
#             topLeft = dfs(n, r, c)
#             topRight = dfs(n, r, c + n)
#             bottomLeft = dfs(n, r + n, c)
#             bottomRight = dfs(n, r + n, c + n)
#             return Node(0, False, topLeft, topRight, bottomLeft, bottomRight)
#         return dfs(len(grid), 0, 0)

"""
T.C.: O(N^2log(N))
這裡的時間複雜度 O(n² log n) 來自於「每一層遞迴都要檢查一個 n x n 的區塊」，而且總共有 log n 層。

 1. 每層遞迴的工作量
在原始寫法中，每次呼叫 dfs 時，會用兩層 for 迴圈檢查這個 n x n 區塊是不是全部一樣。這個檢查需要 O(n²) 時間。
 2. 遞迴分割
如果不是 leaf node，會把這個區塊分成四個 n/2 x n/2 的子區塊，對每個子區塊再遞迴呼叫一次。
 3. 遞迴關係式
所以總的時間複雜度 T(n) 可以寫成：￼這是典型的分治遞迴（Divide and Conquer）。
 4. 主定理（Master Theorem），是一個用來分析分治遞迴時間複雜度的數學工具。
 T(n) = a * T(n/b) + O(n^d)
 其中：
 • a：每次分成幾個子問題
 • n/b：每個子問題的規模
 • O(n^d)：每層合併或分割時所需的額外工作量

主定理的三種情況 ￼
 1. 如果 a < b^d
T(n) = O(n^d)
 2. 如果 a = b^d
T(n) = O(n^d log n)
 3. 如果 a > b^d
T(n) = O(n^{log_b a})

舉例來說，a=4, b=2, d=2：
T(n) = 4 * T(n/2) + O(n^2)
 ▫ a = 4（每次分成 4 個子問題）
 ▫ b = 2（每個子問題的規模是原來的一半）
 ▫ d = 2（合併時的工作量是 O(n²)）
根據主定理，T(n) = O(n² log n)。

S.C.: O(N^2)
"""


class Solution:
    def construct(self, grid: List[List[int]]) -> "Node":
        def dfs(n: int, r: int, c: int) -> Node:
            if n == 1:
                return Node(grid[r][c], True, None, None, None, None)

            n = n // 2
            topLeft = dfs(n, r, c)
            topRight = dfs(n, r, c + n)
            bottomLeft = dfs(n, r + n, c)
            bottomRight = dfs(n, r + n, c + n)

            if (
                topLeft.isLeaf
                and topRight.isLeaf
                and bottomLeft.isLeaf
                and bottomRight.isLeaf
                and topLeft.val == topRight.val == bottomLeft.val == bottomRight.val
            ):
                return Node(topLeft.val, True, None, None, None, None)

            return Node(0, False, topLeft, topRight, bottomLeft, bottomRight)

        return dfs(len(grid), 0, 0)


"""
T.C.: O(N^2)
這個解法的時間複雜度是 O(n²)，因為每個格子只會被遞迴拜訪一次。

 • 這個遞迴函式會把網格不斷分成四個區塊，直到每個 cell 都成為 leaf node。
 • 沒有在每一層用 for 迴圈掃描整個子區塊。
 • 每個 cell 只會被處理一次（建立 leaf node 或合併）。
 • 所以總共的操作次數就是格子的數量：O(n²)。

S.C.: O(N^2)
"""
